### ISSUE TITLE
Define run_opts in hparam.yaml

### ISSUE THREAD

**Issue by larsrgit** [speechbrain/speechbrain#1133]
It would be useful if there was the possibility to define run_opts in the hyperparameter yaml file.

The split of some parameters into run_opts and hparams is not always intuitive (e.g in my opinion the max_grad_norm should be a hyperparamter since the choice of when to clip gradients is strongly dependent of the model, optimizer, loss and the learning rate). By enabling the user to set the run_opts in the hyperparameter yaml file, the user would not need to handle run-opts and hyperparameters differently.

Currently I just overwrite all run-opts that exist in hparams after loading the yaml. But this disables the ability to change those parameters with overwrites in the command line. (run-opts are not returned as overrides by sb.parse_arguments(), so when overriding the run-opts if they exist in hparams, I can't distinguish if those run-opts where set by default or in the command line).

**Comment by Gastron** [comment]
Actually, run_opts can be set via hparams as well. Looks like this isn't documented.

https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/core.py#L437-L453

**Comment by larsrgit** [comment]
Oh this is nice. But I think this does not work for run_opts that have a default in speech_brain.parse_arguments, because they will always be in the run_opts passed to the brain initialization. 
Maybe there should not be two places in the code with defaults for the run_opts?

**Comment by Gastron** [comment]
Right, we should maybe remove the defaults from the SpeechBrain arguments parsing, and keep the defaults in Brain. We could also have it read the command line if no args are given. I'll make a PR quick.

**Comment by larsrgit** [comment]
Okay, I can understand the reasoning for having two places for defaults.
I still think the priority should always be command line > hparam-file > defaults. The way it is now, for run_opts with a default in parse_args the priority is command line > defaults and the hparam file is ignored.

But it's no big problem for me since there currently is not many defaults in parse-args I want to override in the hparam yaml, so I will just handle them manually.

**Comment by Gastron** [comment]
I agree, that would be the sane priority order, but it's not possible with the current implementation. 

We could maybe load the YAML in the argument parsing though, hmm, gotta think about this.

**Comment by larsrgit** [comment]
If somehow argument parsing would return the information about which run-opts where set in the command line and which by defaults, we could overwrite those which were set by default by the hparams loaded from the yaml.

**Comment by Gastron** [comment]
Yeah, but additionally some run_opts parameters need to be known right after the argument parsing, so the hparams options should be injected inside the argument parsing function. That might actually be a reasonable approach anyway, but it would break backwards compatibility. I think this is something to consider for the next major release. 

In the meanwhile, I think the run_opts that people might realistically want to specify in YAML are already available: max_grad_norm, non_finite_patience, jit_module_keys, ckpt_interval_minutes. Adding auto_mix_prec there would be good probably.

**Comment by Gastron** [comment]
Ha, auto_mix_prec via hparams is actually a feature that's being hacked in some recipes already:
https://github.com/speechbrain/speechbrain/blob/34159b84c42715e98335a961d5eb909dd1f29985/recipes/WSJ0Mix/separation/train.py#L517

**Comment by larsrgit** [comment]
yeah, thats the way I'm doing it right now. No big deal (since I know now I only have to do it for run_opts which have a default), but may still be good to change in the future.

**Comment by Gastron** [comment]
With these PRs merged, I'll close the issue, I think further changes should be considered only as part of some bigger refactor.
